test/core/end2end/inproc_callback_test.cc

```diff
@@ -0,0 +1,498 @@
+                                            grpc_channel_args* server_args) {
+  grpc_end2end_test_fixture f;
+  gpr_log(GPR_INFO, "Running test: %s/%s", test_name, config.name);
+  f = config.create_fixture(client_args, server_args);
+  config.init_server(&f, server_args);
+  config.init_client(&f, client_args);
+  return f;
+}
+
+static gpr_timespec n_seconds_from_now(int n) {
+  return grpc_timeout_seconds_to_deadline(n);
+}
+
+static gpr_timespec five_seconds_from_now() { return n_seconds_from_now(5); }
+
+static void drain_cq(grpc_completion_queue* cq) {
+  // Wait for the shutdown callback to arrive, or fail the test
+  GPR_ASSERT(g_shutdown_callback->Wait(five_seconds_from_now()));
+  gpr_log(GPR_DEBUG, "CQ shutdown wait complete");
+  grpc_core::Delete(g_shutdown_callback);
+}
+
+static void shutdown_server(grpc_end2end_test_fixture* f) {
+  if (!f->server) return;
+  grpc_server_shutdown_and_notify(
+      f->server, f->shutdown_cq,
+      reinterpret_cast<void*>(static_cast<intptr_t>(1000)));
+  GPR_ASSERT(
+      grpc_completion_queue_pluck(f->shutdown_cq, (void*)((intptr_t)1000),
+                                  grpc_timeout_seconds_to_deadline(5), nullptr)
+          .type == GRPC_OP_COMPLETE);
+  grpc_server_destroy(f->server);
+  f->server = nullptr;
+}
+
+static void shutdown_client(grpc_end2end_test_fixture* f) {
+  if (!f->client) return;
+  grpc_channel_destroy(f->client);
+  f->client = nullptr;
+}
+
+static void end_test(grpc_end2end_test_fixture* f) {
+  shutdown_server(f);
+  shutdown_client(f);
+
+  grpc_completion_queue_shutdown(f->cq);
+  drain_cq(f->cq);
+  grpc_completion_queue_destroy(f->cq);
+  grpc_completion_queue_destroy(f->shutdown_cq);
+}
+
+static void simple_request_body(grpc_end2end_test_config config,
+                                grpc_end2end_test_fixture f) {
+  grpc_call* c;
+  grpc_call* s;
+  grpc_op ops[6];
+  grpc_op* op;
+  grpc_metadata_array initial_metadata_recv;
+  grpc_metadata_array trailing_metadata_recv;
+  grpc_metadata_array request_metadata_recv;
+  grpc_call_details call_details;
+  grpc_status_code status;
+  const char* error_string;
+  grpc_call_error error;
+  grpc_slice details;
+  int was_cancelled = 2;
+  char* peer;
+  gpr_timespec deadline = five_seconds_from_now();
+
+  c = grpc_channel_create_call(f.client, nullptr, GRPC_PROPAGATE_DEFAULTS, f.cq,
+                               grpc_slice_from_static_string("/foo"), nullptr,
+                               deadline, nullptr);
+  GPR_ASSERT(c);
+
+  peer = grpc_call_get_peer(c);
+  GPR_ASSERT(peer != nullptr);
+  gpr_log(GPR_DEBUG, "client_peer_before_call=%s", peer);
+  gpr_free(peer);
+
+  grpc_metadata_array_init(&initial_metadata_recv);
+  grpc_metadata_array_init(&trailing_metadata_recv);
+  grpc_metadata_array_init(&request_metadata_recv);
+  grpc_call_details_init(&call_details);
+
+  // Create a basic client unary request batch (no payload)
+  memset(ops, 0, sizeof(ops));
```

Comment:
I'm not sure if this is very trivial but it may be helpful to have a small helper function to provide this op array. Also if these ops are common among multiple tests it may be useful to have one common function in a library to return an array of ops that may be consumed by those tests.

Reply:
Great observation... but we try not to reuse code or have helper functions in tests for two reasons.
1. Readability - let it be immediately obvious what each test is trying to do
1. Testability - every helper function itself needs to get tested

So there are cases where we use helpers when the overall code is incredibly complex, but in those cases we try to have tests for the helpers too.